File: 0514-freedom-trail/0514-freedom-trail.py
class Solution:
    def findRotateSteps(self, ring: str, key: str) -> int:
        MOD = n = len(ring)
        
        def neighbors(dial):
            for step in range(n):
                anti_clock = (dial - step) % MOD
                clock = (dial + step) % MOD
                for ndial in (anti_clock, clock):
                    yield step, ndial
        
        @cache
        def dfs(dial, index):
            if index >= len(key):
                return 0
            
            steps = inf
            
            for step, ndial in neighbors(dial):
                if ring[ndial] == key[index]:
                    steps = min(steps, step + dfs(ndial, index+1))
            
            return steps
        
        return len(key) + dfs(0, 0)

File: 0514-freedom-trail/0514-freedom-trail.java
class Solution {
    public int findRotateSteps(String ring, String key) {
        return dfs(ring, key, 0, 0, new Integer[ring.length()][key.length()]);
    }
    
     public int dfs(String ring, String key, int ringIndex, int keyIndex, Integer[][] dp) {
        if (keyIndex == key.length()) {
            return 0;
        }

        if (dp[ringIndex][keyIndex] != null) {
            return dp[ringIndex][keyIndex];
        }

        int minSteps = Integer.MAX_VALUE;

        for (int i = 0; i < ring.length(); i++) {
            if (ring.charAt(i) == key.charAt(keyIndex)) {

                int clockWise = Math.abs(ringIndex - i);
                int antiClockWise = ring.length() - clockWise;

                int steps = 1 + Math.min(clockWise, antiClockWise) 
                    + dfs(ring, key, i, keyIndex + 1, dp);

                minSteps = Math.min(minSteps, steps);
            }
        }

        dp[ringIndex][keyIndex] = minSteps;

        return minSteps;
    }
}